adlfs/transfer.py

```diff
@@ -0,0 +1,368 @@
+import logging
+import multiprocessing
+import os
+import pickle
+import threading
+import time
+import uuid
+
+from adlfs.utils import clamp
+
+MAX_THREADS = 64
+
+logger = logging.getLogger(__name__)
+
+File = namedtuple('File', 'src dst state nbytes start stop chunks')
+Chunk = namedtuple('Chunk', 'name state offset retries')
+
+
+class DisjointState(object):
+    """
+    Collection of disjoint sets containing state about objects.
+
+    Internal class used by `ADLTransferClient`.
+    """
+    def __init__(self, *states):
+        self._states = {state: set() for state in states}
+        self._objects = {}
+
+    @property
+    def states(self):
+        return list(self._states)
+
+    @property
+    def objects(self):
+        return list(self._objects)
+
+    def __iter__(self):
+        return iter(self._objects.items())
+
+    def __getitem__(self, obj):
+        return self._objects[obj]
+
+    def __setitem__(self, obj, state):
+        if obj in self._objects:
+            self._states[self._objects[obj]].discard(obj)
+        self._states[state].add(obj)
+        self._objects[obj] = state
+
+    def contains_all(self, state):
+        objs = self._states[state]
+        return len(objs) > 0 and len(self.objects) - len(objs) == 0
+
+    def contains_none(self, *states):
+        return all([len(self._states[state]) == 0 for state in states])
+
+    def __str__(self):
+        status = " ".join(
+            ["%s=%d" % (s, len(self._states[s])) for s in self._states])
+        return "<DisjointState: " + status + ">"
+
+    __repr__ = __str__
+
+
+class ADLTransferClient(object):
+    """
+    Client for transferring data from/to Azure DataLake Store
+
+    Parameters
+    ----------
+    adlfs: ADL filesystem instance
+    name: str
+        Unique ID used for persistence.
+    transfer: callable
+        Function or callable object invoked when transferring chunks. See
+        ``Function Signatures``.
+    merge: callable [None]
+        Function or callable object invoked when merging chunks. If None,
+        then merging is skipped. See ``Function Signatures``.
+    nthreads: int [None]
+        Number of threads to use (range is [1, MAX_THREADS]). If None, uses the
+        number of cores.
+    chunksize: int [2**26]
+        Number of bytes in each chunk for splitting big files. Files smaller
+        than this number will always be transferred in a single thread.
+    tmp_path: str ['/tmp']
+        Path used for temporarily storing transferred chunks until chunks
+        are gathered into a single file. If None, then each chunk will be
+        written into the same file.
+    tmp_prefix: str ['part_']
+        If given and not None, this is used to provide a prefix to the
+        temporary chunk.
+    tmp_unique: bool [True]
+        If True, then a unique ID will be generated to create a subdirectory
+        containing the temporary chunks. Otherwise, all temporary chunks
+        will be placed in `tmp_path`.
```

Comment:
Ah, I see. Can you give an example of what this looks like for a given file? Will the caller be able to identify what file this unique tmp corresponds to? It should probably also include the target file name somewhere in the path for identification by the caller in the event of failure.